test/cpp/end2end/xds/xds_end2end_test.cc

```diff
@@ -2975,6 +2815,206 @@ TEST_P(XdsRbacTestWithActionPermutations, AnyPermissionOrIdPrincipal) {
+  backends_[0]->Start();
+  backends_[0]->notifier()->WaitOnServingStatusChange(
+      absl::StrCat(ipv6_only_ ? "[::1]:" : "127.0.0.1:", backends_[0]->port()),
+      grpc::StatusCode::OK);
+  SendRpc([this]() { return CreateInsecureChannel(); }, {}, {},
+          /*test_expects_failure=*/GetParam().rbac_action() == RBAC_Action_DENY,
+          grpc::StatusCode::PERMISSION_DENIED);
+  // Only the second rbac logs if it denies the rpc.
+  if (GetParam().rbac_action() == RBAC_Action_DENY) {
+    EXPECT_THAT(audit_logs_,
+                ::testing::ElementsAreArray(
+                    {"{\"authorized\":false,\"matched_rule\":\"policy\","
+                     "\"policy_name\":\"rbac2\",\"principal\":\"\",\"rpc_"
+                     "method\":\"/grpc.testing.EchoTestService/Echo\"}"}));
+  } else {
+    EXPECT_THAT(audit_logs_, ::testing::ElementsAre());
+  }
+}
+
+TEST_P(XdsRbacTestWithActionPermutations,
+       MultipleRbacPoliciesWithAuditOnDenyAndAllow) {
+  ScopedExperimentalEnvVar env_var("GRPC_EXPERIMENTAL_XDS_RBAC_AUDIT_LOGGING");
+  RBAC always_allow;
+  auto* rules = always_allow.mutable_rules();
+  rules->set_action(RBAC_Action_ALLOW);
+  Policy policy;
+  policy.add_permissions()->set_any(true);
+  policy.add_principals()->set_any(true);
+  (*rules->mutable_policies())["policy"] = policy;
+  auto* logging_options = rules->mutable_audit_logging_options();
+  logging_options->set_audit_condition(
+      RBAC_AuditLoggingOptions_AuditCondition_ON_DENY_AND_ALLOW);
+  auto* audit_logger =
+      logging_options->add_logger_configs()->mutable_audit_logger();
+  audit_logger->mutable_typed_config()->set_type_url("/test_logger");
+  TypedStruct typed_struct;
+  typed_struct.set_type_url("/test_logger");
+  typed_struct.mutable_value()->mutable_fields();
+  audit_logger->mutable_typed_config()->PackFrom(typed_struct);
+  RBAC rbac;
+  rules = rbac.mutable_rules();
+  rules->set_action(GetParam().rbac_action());
+  (*rules->mutable_policies())["policy"] = policy;
+  logging_options = rules->mutable_audit_logging_options();
+  logging_options->set_audit_condition(
+      RBAC_AuditLoggingOptions_AuditCondition_ON_DENY_AND_ALLOW);
+  audit_logger = logging_options->add_logger_configs()->mutable_audit_logger();
+  audit_logger->mutable_typed_config()->PackFrom(typed_struct);
+  SetServerRbacPolicies(default_server_listener_,
+                        {always_allow, rbac, always_allow});
+  backends_[0]->Start();
+  backends_[0]->notifier()->WaitOnServingStatusChange(
+      absl::StrCat(ipv6_only_ ? "[::1]:" : "127.0.0.1:", backends_[0]->port()),
+      grpc::StatusCode::OK);
+  SendRpc([this]() { return CreateInsecureChannel(); }, {}, {},
+          /*test_expects_failure=*/GetParam().rbac_action() == RBAC_Action_DENY,
+          grpc::StatusCode::PERMISSION_DENIED);
+  // If the second rbac denies the request, the last rbac won't log. Otherwise
+  // all rbacs log.
+  if (GetParam().rbac_action() == RBAC_Action_DENY) {
```

Comment:
When I suggested using `ElementsAreArray()`, I meant something like this:

```
std::vector<absl::string_view> expected = {
    "{\"authorized\":true,\"matched_rule\":\"policy\","
    "\"policy_name\":\"rbac1\",\"principal\":\"\",\"rpc_"
    "method\":\"/grpc.testing.EchoTestService/Echo\"}"
};
if (GetParam().rbac_action() == RBAC_Action_DENY) {
  expected.push_back(
      "{\"authorized\":false,\"matched_rule\":\"policy\","
      "\"policy_name\":\"rbac1\",\"principal\":\"\",\"rpc_"
      "method\":\"/grpc.testing.EchoTestService/Echo\"}");
} else {
  expected.push_back(
      "{\"authorized\":true,\"matched_rule\":\"policy\","
      "\"policy_name\":\"rbac1\",\"principal\":\"\",\"rpc_"
      "method\":\"/grpc.testing.EchoTestService/Echo\"}");
  expected.push_back(
      "{\"authorized\":true,\"matched_rule\":\"policy\","
      "\"policy_name\":\"rbac1\",\"principal\":\"\",\"rpc_"
      "method\":\"/grpc.testing.EchoTestService/Echo\"}");
}
EXPECT_THAT(audit_logs_, ElementsAreArray(expected));
```

In other words, the advantage of `ElementsAreArray()` over `ElementsAre()` is that the former allows you to construct the list separately, not specifying it as a literal in the expectation itself.  If you're going to specify it linearly in the expectation itself, then you might as well stick with `ElementsAre()`, because `ElementsAreArray()` doesn't actually provide any additional value (in fact, it's actually slightly more cumbersome).

Having said that, though, it occurs to me that in this case, it's probably easier to just say this:

```
constexpr absl::string_view kAllowedLog =
    "{\"authorized\":true,\"matched_rule\":\"policy\","
    "\"policy_name\":\"rbac1\",\"principal\":\"\",\"rpc_"
    "method\":\"/grpc.testing.EchoTestService/Echo\"}";
constexpr absl::string_view kDeniedLog =
    "{\"authorized\":false,\"matched_rule\":\"policy\","
    "\"policy_name\":\"rbac1\",\"principal\":\"\",\"rpc_"
    "method\":\"/grpc.testing.EchoTestService/Echo\"}";
if (GetParam().rbac_action() == RBAC_Action_DENY) {
  EXPECT_THAT(audit_logs_, ElementsAre(kAllowedLog, kDeniedLog));
} else {
  EXPECT_THAT(audit_logs_, ElementsAre(kAllowedLog, kAllowedLog, kAllowedLog));
}
```

Reply:
Ah I didn't realize and indeed was a bit confused about what advantage `ElementsAreArray` could have.

So I took your first suggestion because the policy name (the filter name here) is not the same for the first and third allowed log. Three rbacs have different names.